quantum/process_keycode/process_swap_hands.c

```diff
@@ -0,0 +1,150 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "process_swap_hands.h"
+#include "quantum_keycodes.h"
+#include "action_tapping.h"
+#include "matrix.h"
+
+extern const keypos_t PROGMEM hand_swap_config[MATRIX_ROWS][MATRIX_COLS];
+
+static matrix_row_t swap_state[MATRIX_ROWS];
+
+static bool     swap_hands_enabled = false;
+static bool     oneshot_enabled    = false;
+static uint16_t oneshot_expiry     = 0;
+
+#if defined(ONESHOT_TIMEOUT) && (ONESHOT_TIMEOUT > 0)
+static inline bool has_oneshot_swaphands_timed_out(void) {
+    return (oneshot_expiry && timer_expired(timer_read(), oneshot_expiry));
+#endif
+}
+
+static inline void swap_hands_task(void) {
+#if defined(ONESHOT_TIMEOUT) && (ONESHOT_TIMEOUT > 0)
+    if (has_oneshot_swaphands_timed_out()) {
+        clear_oneshot_swaphands();
+    }
+#endif
+}
+
+void clear_oneshot_swaphands(void) {
+    oneshot_expiry     = 0;
+    swap_hands_enabled = false;
+    oneshot_enabled    = false;
+}
+
+void preprocess_swap_hands(keyrecord_t *record) {
+    // TODO: implement a "real" swap_hands_task called from matrix_scan_quantum?
+    swap_hands_task();
+
+    keypos_t     pos     = record->event.key;
+    matrix_row_t col_bit = MATRIX_ROW_SHIFTER << pos.col;
+    bool         do_swap = record->event.pressed ? swap_hands_enabled : swap_state[pos.row] & (col_bit);
+
+    if (do_swap) {
+        record->event.key.row = pgm_read_byte(&hand_swap_config[pos.row][pos.col].row);
+        record->event.key.col = pgm_read_byte(&hand_swap_config[pos.row][pos.col].col);
+        swap_state[pos.row] |= col_bit;
+    } else {
+        swap_state[pos.row] &= ~(col_bit);
+    }
+}
+
+bool process_swap_hands(uint16_t keycode, keyrecord_t *record) {
+    static uint16_t last_keycode = 0;
+
+    bool pressed = record->event.pressed;
+
+    uint16_t previous_keycode = last_keycode;
+    last_keycode              = keycode;
+
+    bool interupted = (previous_keycode != keycode);
+
+    switch (keycode) {
+        case SH_TG:
+            if (pressed) {
+                swap_hands_enabled = !swap_hands_enabled;
+            }
+            break;
+        case SH_MON:
+            swap_hands_enabled = pressed;
+            break;
+        case SH_MOFF:
+            swap_hands_enabled = !pressed;
+            break;
+        case SH_ON:
+            if (!pressed) {
+                swap_hands_enabled = true;
+            }
+            break;
+        case SH_OFF:
+            if (!pressed) {
+                swap_hands_enabled = false;
+            }
+            break;
+        case SH_TT: {
```

Comment:
Logic for tap is broken here.  holding will work, init tap will work.  But after the tap, to get out of swap hands mode, you have to hold, hit another key and then release. 

I suspect that this is because a lot of the tap detection is handled by action.c/action_tapping.c, and that code has been removed.

Reply:
Is `SH_TT` with `TAPPING_TOGGLE 1` any different than just `SH_MON`?

Reply:
meant to get back to you on this. 

Yes.  If held past tapping term, it functions as momentary, but if it's tapped (eg, not held past tapping term) it toggles the status.  

TG would only toggle/invert the status, and doesn't have the MO type functionality.